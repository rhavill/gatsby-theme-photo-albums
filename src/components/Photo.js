import React from "react";
import PropTypes from "prop-types"

import Header from './Header';

export default ({location}) => {
  console.log('Photo location', location)
  return (
    <div>
      <Header path={location.pathname}/>
      <div>photo</div>
    </div>
  )
}

Header.propTypes = {
  location: PropTypes.string.isRequired,
}