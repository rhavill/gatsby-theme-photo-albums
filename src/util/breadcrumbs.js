import toTitleCase from './to-title-case';

const breadcrumbs = (path, rootTitle = 'Home') => {
  // Set path to empty string for root directory. 
  // Also, remove trailing slash (if it exists) for consistency.
  const parts = path === '/' ? [''] : path.replace(/\/$/, '').split('/')
  return parts
    .map((folder, i) => ({
      // convert to "title case" and remove file extension
      title: folder === '' ? rootTitle : toTitleCase(folder).replace(/\.[^/.]+$/, ''),
      path: folder === '' ? '/' : parts.filter((part, j) => j <= i).join('/') 
    }))
}

export default breadcrumbs