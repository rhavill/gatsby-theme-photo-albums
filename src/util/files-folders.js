const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const match = require('ramda/src/match')

const getChildPaths = curry(
  (currentPath, paths) => {
    return filter(isParentDirectoryOf(currentPath), paths)
  }
)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

module.exports = getChildPaths