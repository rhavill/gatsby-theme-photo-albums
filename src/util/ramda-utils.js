const map = require('ramda/src/map')
const prop = require('ramda/src/prop')

/**
 * Accepts an array of objects and returns an array containing the values of a 
 * single property from each object of the input array.
 *
 * @func
 * @sig Array -> Array
 * @param {String} property The property to extract from each object of the 
 *    input array.
 * @param {Array} objects The input array containing objects.
 * @return {Array} A simple array containing values exracted from the `property` 
 *    values of the input array.
 * @example
 *
 *      const objects = [{name: 'Hilda', age: 33}, {name: 'José', age: 45}]
 *      objectArrayToPropArray('name', objects); //=> ['Hilda', 'José']
 */
exports.objectArrayToPropArray =  (property, objects) => map(prop(property), objects)
