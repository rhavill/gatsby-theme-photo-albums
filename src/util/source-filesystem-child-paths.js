const compose = require('ramda/src/compose')
const concat = require('ramda/src/concat')
const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const map = require('ramda/src/map')
const match = require('ramda/src/match')
const replace = require('ramda/src/replace')

/**
 * When provided with a Gatsby location path and some gatsby-source-filesystem 
 * relativePath values, determines the Gatsby-compatible location paths of the 
 * child directories.
 * @param {string} currentPath The Gatsby location path of a possible parent directory. 
 * @param {string[]} folders An array of gatsby-source-filesystem relativePath values from the allDirectory query type.
 * @returns {string[]} An array of Gatsby location paths that are children of currentPath
 */
const getChildPaths = curry(
  (currentPath, folders) => 
    compose(
      filterChildPaths(currentPath),
      addLeadingSlashes
    )(folders)
)

const filterChildPaths = currentPath =>
  filter(
    compose(
      isParentDirectoryOf,
      removeTrailingSlashAndPageNumber
    )(currentPath)
  )

// remove trailing slash and digit if it they exist.
const removeTrailingSlashAndPageNumber = replace(/\/\d+$/, '')

const addLeadingSlash = concat('/')

// Add leading slashes to filesystem paths to be consistent with Gatsby location path
const addLeadingSlashes = map(addLeadingSlash)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

module.exports = getChildPaths