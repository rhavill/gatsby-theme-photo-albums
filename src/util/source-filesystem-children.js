const match = require('ramda/src/match')

const getChildren = (folder, data) => {
  // remove trailing slash if it exists
  const currentPath = folder.replace(/\/\d+$/, '')
  const folders = [], files = []
  data.allDirectory.edges.forEach(edge => {
    // Add leading slash to filesystem path to be consistent with current location path
    const fsPath = '/' + edge.node.relativePath
    if (isParentDirectoryOf(currentPath, fsPath)) {
      folders.push(fsPath)
    }
  })
  data.allFile.edges.forEach(edge => {
    // Add leading slash to filesystem file and directory to be consistent with current location path
    const fsPath = '/' + edge.node.relativePath
    const fixed = edge.node.childImageSharp.fixed;
    if (isParentDirectoryOf(currentPath, fsPath)) {
      files.push({fsPath, fixed})
    }
  })
  const children = {
    folders,
    files
  };
  return children;
}

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = (parent, child) => {
  const regex = new RegExp('^' + parent + '/?[^/]+$')
  return match(regex, child).length > 0
}

module.exports = {
  getChildren,
  isParentDirectoryOf,
}