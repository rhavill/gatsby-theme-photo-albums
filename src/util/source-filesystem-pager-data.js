const getChildPaths = require('./source-filesystem-child-paths')

/**
 * When provided with a gatsby-source-filesystem relativePath of a directory 
 * and some gatsby-source-filesystem relativePath values of files, determines 
 * pager data needed to display the files in multiple pages.
 * @param {string} currentPath The gatsby-source-filesystem relativePath of a possible parent directory. 
 * @param {string[]} files An array of gatsby-source-filesystem relativePath values of files.
 * @param {integer} filesPerPage The maximum number of files to display on each page
 * @returns {object[]} An array of pager data objects for each child file of the current path
 */
const getPagerData = (currentPath, files, filesPerPage) => {
  const pagerData = []
  const children = getChildPaths(`/${currentPath}`, files)
  const childCount = children.length
  const numPages = Math.ceil(childCount / filesPerPage) || 1
  Array.from({ length: numPages }).forEach((_, i) => {
    pagerData.push({
      limit: filesPerPage,
      skip: i * filesPerPage,
      numPages,
      currentPage: i + 1
    })
  })
  return pagerData
}

module.exports = {
  getPagerData
}