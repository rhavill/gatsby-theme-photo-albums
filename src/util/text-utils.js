import compose from 'ramda/src/compose'
import path from 'ramda/src/path'
import replace from 'ramda/src/replace'

const toTitleCase = text => 
  text.toLowerCase()
    .split(/[-_]+/)
    .map((s) => s.charAt(0).toUpperCase() + s.substring(1))
    .join(' ')

const removeFileExtension = replace(/^(.+)\.[^.]+$/, '$1')

const pathToFile = replace(/.*\/([^/]+$)/, '$1')

const pathToFileTitle = compose(toTitleCase, removeFileExtension, pathToFile)

const removePathPrefix = data => replace(
  path(['site', 'pathPrefix'], data), 
  ''
)

const gatsbyPathnameToChildComponentPath = (pathname, graphqlData) =>
  compose(removePathPrefix(graphqlData), decodeURIComponent)(pathname)

export {
  pathToFile,
  pathToFileTitle,
  removeFileExtension,
  toTitleCase,
  gatsbyPathnameToChildComponentPath,
}