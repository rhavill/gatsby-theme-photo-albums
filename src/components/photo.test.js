import React from 'react'
import renderer from 'react-test-renderer'
import {cleanup, render} from '@testing-library/react'
import Photo from '../pages/photo'
import photoData from '../test-data/source-filesystem-photo-data'
import {pathToFileTitle} from '../util/text-utils'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

describe('Photo', () => {
  it('renders correctly', () => {
    const pageContext = {basePath: '/'}
    const tree = renderer
      .create(<Photo path='1.jpg' data={photoData.data} pageContext={pageContext}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('displays images with "title" attributes containing a title-case version of the photo file name', () => {
    const pageContext = {basePath: '/'}
    const path = photoData.data.photo.childImageSharp.fluid.src
    const {getAllByTitle} = render(
      <Photo path={path} data={photoData.data} pageContext={pageContext}/>
    )
    const expected = pathToFileTitle(path)
    expect(getAllByTitle(expected).length).toBe(2)
  })
  it('passes decoded url to child components', () => {
    const pageContext = {basePath: '/'}
    const path = '/san-sebasti%c3%a1n/san-sebasti%c3%a1n.jpg'  
    const {getAllByTestId} = render(
      <Photo path={path} data={photoData.data} pageContext={pageContext}/>
    )
    expect(getAllByTestId('/san-sebastián/san-sebastián.jpg').length).toBe(1)
  })
  it('passes path to child components without pathPrefix config variable', () => {
    const pageContext = {basePath: '/'}
    const path = '/path-prefix/top-level/file.jpg'  
    const {getAllByTestId} = render(
      <Photo path={path} data={photoData.data} pageContext={pageContext}/>
    )
    expect(getAllByTestId('/top-level/file.jpg').length).toBe(1)
  })
})