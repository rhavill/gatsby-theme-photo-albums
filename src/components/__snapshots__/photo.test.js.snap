// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo renders correctly 1`] = `
<div
  className="wrapper"
>
  <header>
    <nav
      className="container"
    >
      <a
        href="1.jpg"
      >
        1
      </a>
    </nav>
  </header>
  <div
    className="container page"
  >
    <div
      className="photo-page"
      data-testid="1.jpg"
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "overflow": "hidden",
            "position": "relative",
          }
        }
      >
        <div
          style={
            Object {
              "paddingBottom": "75%",
              "width": "100%",
            }
          }
        />
        <img
          alt=""
          src="data:image/jpeg;base64,/9j/2wBDABALDA4MChAODQ4SERATGCgaGBYWGDEjJR0oOjM9PDkzODdASFxOQERXRTc4UG1RV19iZ2hnPk1xeXBkeFxlZ2P/2wBDARESEhgVGC8aGi9jQjhCY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2P/wgARCAAPABQDASIAAhEBAxEB/8QAGAAAAwEBAAAAAAAAAAAAAAAAAAQFAQP/xAAVAQEBAAAAAAAAAAAAAAAAAAABAP/aAAwDAQACEAMQAAABxyT0CiIDf//EABwQAAIBBQEAAAAAAAAAAAAAAAACAQMEEBESE//aAAgBAQABBQKlcHqrEuuOjcn/xAAUEQEAAAAAAAAAAAAAAAAAAAAQ/9oACAEDAQE/AT//xAAVEQEBAAAAAAAAAAAAAAAAAAAQEf/aAAgBAgEBPwGn/8QAGhAAAQUBAAAAAAAAAAAAAAAAAAECEBExkf/aAAgBAQAGPwKn9mzVNP/EABsQAQACAwEBAAAAAAAAAAAAAAEAESExURCB/9oACAEBAAE/IWNSdTqK4S1i40YGpY0H3wf/2gAMAwEAAgADAAAAEAzP/8QAFREBAQAAAAAAAAAAAAAAAAAAEQD/2gAIAQMBAT8QCL//xAAWEQEBAQAAAAAAAAAAAAAAAAAAEQH/2gAIAQIBAT8Qylf/xAAbEAEBAQEAAwEAAAAAAAAAAAABEQBBITFRcf/aAAgBAQABPxCSEdrGnH7vMEV7XmDCnkcIEVXCkAIo1DW7f0Lv/9k="
          style={
            Object {
              "height": "100%",
              "left": 0,
              "objectFit": "cover",
              "objectPosition": "center",
              "opacity": 1,
              "position": "absolute",
              "top": 0,
              "transitionDelay": "500ms",
              "width": "100%",
            }
          }
          title="1"
        />
        <picture>
          <source
            sizes="(max-width: 1024px) 100vw, 1024px"
            srcSet="/static/d48e4ab75840455d20bbbb33140bcdaa/160a1/IMG_20190801_180122560_HDR-small.jpg 512w,
/static/d48e4ab75840455d20bbbb33140bcdaa/4271c/IMG_20190801_180122560_HDR-small.jpg 614w,
/static/d48e4ab75840455d20bbbb33140bcdaa/42a5a/IMG_20190801_180122560_HDR-small.jpg 819w,
/static/d48e4ab75840455d20bbbb33140bcdaa/426ce/IMG_20190801_180122560_HDR-small.jpg 1024w"
          />
          <img
            alt="1"
            loading="lazy"
            onLoad={[Function]}
            sizes="(max-width: 1024px) 100vw, 1024px"
            src="/static/d48e4ab75840455d20bbbb33140bcdaa/426ce/IMG_20190801_180122560_HDR-small.jpg"
            srcSet="/static/d48e4ab75840455d20bbbb33140bcdaa/160a1/IMG_20190801_180122560_HDR-small.jpg 512w,
/static/d48e4ab75840455d20bbbb33140bcdaa/4271c/IMG_20190801_180122560_HDR-small.jpg 614w,
/static/d48e4ab75840455d20bbbb33140bcdaa/42a5a/IMG_20190801_180122560_HDR-small.jpg 819w,
/static/d48e4ab75840455d20bbbb33140bcdaa/426ce/IMG_20190801_180122560_HDR-small.jpg 1024w"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
            title="1"
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"/static/d48e4ab75840455d20bbbb33140bcdaa/160a1/IMG_20190801_180122560_HDR-small.jpg 512w,
/static/d48e4ab75840455d20bbbb33140bcdaa/4271c/IMG_20190801_180122560_HDR-small.jpg 614w,
/static/d48e4ab75840455d20bbbb33140bcdaa/42a5a/IMG_20190801_180122560_HDR-small.jpg 819w,
/static/d48e4ab75840455d20bbbb33140bcdaa/426ce/IMG_20190801_180122560_HDR-small.jpg 1024w\\" sizes=\\"(max-width: 1024px) 100vw, 1024px\\" /><img loading=\\"lazy\\" sizes=\\"(max-width: 1024px) 100vw, 1024px\\" srcset=\\"/static/d48e4ab75840455d20bbbb33140bcdaa/160a1/IMG_20190801_180122560_HDR-small.jpg 512w,
/static/d48e4ab75840455d20bbbb33140bcdaa/4271c/IMG_20190801_180122560_HDR-small.jpg 614w,
/static/d48e4ab75840455d20bbbb33140bcdaa/42a5a/IMG_20190801_180122560_HDR-small.jpg 819w,
/static/d48e4ab75840455d20bbbb33140bcdaa/426ce/IMG_20190801_180122560_HDR-small.jpg 1024w\\" src=\\"/static/d48e4ab75840455d20bbbb33140bcdaa/426ce/IMG_20190801_180122560_HDR-small.jpg\\" alt=\\"1\\" title=\\"1\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
